### DISCUSSION TITLE
Active maintainers wanted!

### DISCUSSION THREAD

**Discussion by aleclarson** [radashi-org/radashi#4] [Announcements]
The goal of Radashi is to continue the vision of Radash under new leadership, so it can receive the love and attention it deserves. Early in Radashi's life, backwards compatibility is critical, so Radash users can easily migrate.

If you can afford dedicating some time to reviewing PRs, closing issues, or even contributing your own features and fixes, Radashi ~wants~ **needs you** on its team of maintainers.

### What's expected

As a maintainer, you're expected to uphold the original vision of Radash. Beyond that, it's up to you how much time you will dedicate to Radashi. 

**Ideally, you've submitted pull requests to Radash or Radashi already,** and so you've demonstrated the requisite knowledge and aptitude.

### How do I join?

Comment here, along with any PRs you've submitted or work you've done on your own fork of Radash. Anyone who applies will be added to the team and given a chance to show their worth.

Inactive maintainers will be removed after a grace period of 3 months, but don't worry, you can always rejoin if you find the time to help out in the future.

**Comment by stefaanMLB** [comment]
I'm interested to join the effort.
I did some PR's on the original [https://github.com/rayepps/radash](https://github.com/rayepps) radash site but these haven't been ported to the [sodiray](https://github.com/sodiray/radash) site.
Created my own private copy of radash in meantime.

**Reply by aleclarson** [reply]
I've sent you an invite, @stefaanMLB.

Feel free to open some PRs, issues, or discussions with whatever's on your mind. You can also take a look at the PRs I've submitted already. I've been working my way through some of Radash's issues, but I probably won't have time to address them all in the short term, so it's great to have your help!

**Comment by localusercamp** [comment]
> vision of Radash

Does this include polymorphism prohibition? =)

I think something like this could improve DX pretty much.
```js
sort(users, user => user.id);
sort(users, 'id'); // support keyof
```

**Reply by aleclarson** [reply]
**TL;DR:** It depends. Polymorphism can be good if it means less confusing function names, but it can be bad if it bloats the implementation excessively. Unfortunately, these are subjective measures.

There's an ideal balance between “API surface area” (i.e. how many functions you need to look at before you find the one that fits your needs) and “implementation bloat” (i.e. if you use a function for one of its purposes, how much code remains in the function that is never used?). The trouble with achieving such a balance is that it needs evaluation on a case-by-case basis. In the case of your `sort` example, I would argue the polymorphism is acceptable if the use case is especially popular.

Leaning towards a minimal “API surface area” means not having multiple functions with similar use cases. Programmers are most productive when it's easy to remember which variant of a function does what they need. One way to achieve this is through verbose naming; for example, Lodash's `pick` could be `pickSpecificProperties` while `pickBy` could be `pickPropertiesWithPredicate`. Of course, that is not the JavaScript way. So in this instance, it would be best for `pick` to support both an array of keys and a predicate function (which is what https://github.com/radashi-org/radashi/pull/30 implements).

With AI code completion becoming more popular, it's also a question of which call signature is most likely to be suggested by your IDE. In other words, if the polymorphism's main intent is to make an existing use case more concise, I'd argue its value is likely not worth the implementation bloat. “Time spent typing” is becoming less and less of a concern.

The vision is to first cover the most common use cases, erring toward simplicity and conciseness. If polymorphism adds *new functionality* without bloating the implementation beyond recognition (the extreme case being Lodash), I'd say it could be acceptable.

The great part is that Radashi is even easier to fork via copy-pasting than Radash is, because of its modular nature and internal use of `import {…} from "radashi"`. If you want more polymorphism or less, it should be simple enough to copy-paste the function and add the polymorphism yourself (possibly using other Radashi functions to make your changes easier to implement).

Hope that helps :)

**Reply by localusercamp** [reply]
> In the case of your sort example, I would argue the polymorphism is acceptable if the use case is especially popular.

I think that is what it is - the most popular use case of specifying one property name (which we could easily do in TS) instead of creating a bunch of functions which only return a single property.

Second thing is how do you know what is more popular before implementing it? =D
I think the best answer is look at similar libraries. Laravel, Lodash, Vue are implementing such polymorphism in one way or another and many people use "shorthands" instead of defining small functions.

In my opinion if all the user need is to filter/sort/etc by single a field the most convenient way is to give him the ability to pass a string. If he wants a more complicated logic he always can use function. And that's it.

P.S. Make some function accept array, string, number, function, and so on is awfull. That's why I did not like Lodash.

I'll be glad to be involved into this project. I thinking of make nuxt module for this repo, what do you think?

**Reply by aleclarson** [reply]
> I think that is what it is - the most popular use case of specifying one property name (which we could easily do in TS) instead of creating a bunch of functions which only return a single property.

One thing to consider is how trivial it is to implement this on your own:
```ts
import * as _ from 'radashi'

export const sortByProperty = (array, prop) => _.sort(array, obj => obj[prop])
```
But if this shorthand signature is indeed popular with >50% of Lodash users, then you'd be right that it's still worth including.

I have a PR almost done that will add such functionality to `sort` and similar Radashi functions.

**Reply by aleclarson** [reply]
> P.S. Make some function accept array, string, number, function, and so on is awfull. That's why I did not like Lodash.

Good, we're on the same page there :)

> I'll be glad to be involved into this project.

Cool, Radashi would appreciate another brain on the team! Let me know if you'd like an invite.

> I thinking of make nuxt module for this repo, what do you think?

Yeah, that sounds like a cool idea. Do you wanna create a new discussion about it where we can discuss the idea further?

**Reply by localusercamp** [reply]
> Do you wanna create a new discussion about it where we can discuss the idea further?

Yeah, I'll create one when I get some time

**Reply by localusercamp** [reply]
> Let me know if you'd like an invite

Give it to me! =)

**Comment by stefaanMLB** [comment]
I support @localusercamp.
In my experience, most users are not interested/willing/able to fork a library and keep the fork in sync with the main library.  Most are just looking for functionality and will most likely appreciate the alternative in the example.
@aleclarson I much appreciate the thoughts you put into this.

**Reply by localusercamp** [reply]
> In my experience, most users are not interested/willing/able to fork a library and keep the fork in sync with the main library.

Yeah, fully agree.

**Reply by aleclarson** [reply]
I assume you're referring to what I said below?

> The great part is that Radashi is even easier to fork via copy-pasting than Radash is, because of its modular nature and internal use of `import {…} from "radashi"`. If you want more polymorphism or less, it should be simple enough to copy-paste the function and add the polymorphism yourself (possibly using other Radashi functions to make your changes easier to implement).

The keyword is "copy-pasting". It's very easy to copy a function from Radashi and paste it into your project as its own utility. That means there's no notion of keeping it in sync with upstream.

**Reply by stefaanMLB** [reply]
OK, the user of the library can start out by copy-pasting but soon enough he/she will want to keep the code in sync across his/her projects and end up creating an NPM module anyway (thus a fork of the library).  
Next thing is to keep the shiny new module up-to-date with changes in the parent library
Next thing he/she will try to have his/her improvements integrated into the parent module to avoid the work of keeping the own module up-to-date with the parent.
This is what happened to me and I guess it's also why you created `radashi` because `radash` isn't accepting PR's.

Granted, the copy-paste option is an alternative and will-suited for people who have a limited number of projected to maintain.

**Reply by aleclarson** [reply]
I understand the concern. While we can't meet everyone's needs, we should try to support at least what >50% of the community will use. Of course, it's hard to gauge that figure, so there's some guesswork required.

I will say that even if you copy-paste 10+ functions from Radashi, you can publish your own package with those customized functions without maintaining a fork. Simply add `radashi` to the package's dependencies and the only thing you'd ever need to update is the version number (like any other dependency).

You could even make `radashi` a peer dependency of the package, so the root project is in control of its `radashi` version. Then you only need to update your “pseudo-fork” when changing the behavior of your own code or upgrading to a new Radashi major.

Finally, your package could forward the exports of `radashi` and override the ones you needed to modify. This is all very straight-forward these days.


```ts
export * from 'radashi'

// Override `select` from radashi with my own version.
export * from './array/select'
```

**Reply by stefaanMLB** [reply]
OK, valid point.  I suggest to put this in the radashi documentation.
I use Typescript for 7 years and it never occurred to me to do this.  Others may not know either.

**Comment by Minhir** [comment]
Hi @aleclarson, thanks for your activity to save that nice library!

I love a simplicity of radash API compared to lodash, and I prefer it as util library in my projects. But it still has a big room for improvements.  I'll be happy to help with code reviews and coding, though I don't have too much dedicated time for it.

**Reply by aleclarson** [reply]
Great to hear you're interested in helping out. Given your familiarity with the project and the helpful code reviews you've done, we'd be glad to have you on the core team. No worries about limited time availability - any involvement is valuable. Want me to add you?

**Reply by Minhir** [reply]
Yes, sure!

**Comment by mctrafik** [comment]
Is this project dead then? I see references to lodash parity page which doesn't exist. I can contribute PR reviews for my part.

**Reply by aleclarson** [reply]
I haven't gotten around to setting up the Lodash Parity page just yet.

I uploaded the data here for now: https://github.com/radashi-org/radashi-org.github.io/blob/lodash-parity/public/lodash_parity_rows.csv

If you're interesting in implementing an Astro component for this page, feel free to submit a PR. The idea would be to load the CSV file from inside the Astro component (using a CSV library) and rendering that as a HTML table (maybe there's a library for CSV-to-HTML transformation, idk).

The docs *website* repository is here: https://github.com/radashi-org/radashi-org.github.io/tree/main

**Comment by Kinbaum** [comment]
@aleclarson I'd like to become a maintainer #13

**Reply by aleclarson** [reply]
Hey! The repo's been very quiet since the AI coding revolution picked up. Less of a need for utility libraries. So I don't think we need more maintainers right now, but more contributors would be great.
